core/src/main/scala/org/bykn/bosatsu/UnusedLetCheck.scala

```diff
@@ -0,0 +1,80 @@
+package org.bykn.bosatsu
+
+import cats.Applicative
+import cats.data.{Chain, Validated, ValidatedNec, Writer, NonEmptyChain}
+import cats.implicits._
+
+import Expr._
+import Identifier.Bindable
+
+object UnusedLetCheck {
+
+  private[this] val ap = Applicative[Writer[Chain[(Bindable, Region)], ?]]
+  private[this] val empty: Writer[Chain[(Bindable, Region)], Set[Bindable]] = ap.pure(Set.empty)
+
+  private[this] def checkArg(arg: Bindable, reg: => Region, w: Writer[Chain[(Bindable, Region)], Set[Bindable]]) =
+    w.flatMap { free =>
+      if (free(arg)) ap.pure(free - arg)
```

Comment:
This is not correctly handling shadowing. Just because you are binding X here, it doesn't mean it wasn't already bound. So, `free - arg` isn't correct, indeed, we unless we are tracking top down scope at the same time, this won't work.

Reply:
we need to add a test that exercises this failing to catch an unused variable, and show that it is broken before continuing. Because on second though, since we are returning the free variables, we do know for sure that arg isn't free if we are binding it, so this may actually be correct.